File: src/application/ROM.java
package application;

import application.excel.Excel;
import application.models.lists.Lists;
import application.models.lists.states.DataState;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardOpenOption;

/**
 * Created by Seth on 4/3/2017.
 */
public class ROM {


	private static byte[] bytes;
	private static int header;
	public static boolean showEmptyValues = false;
	public static boolean showMonsterDuplicates = false;
	private static File programDirectory;
	private static File rom;
	private static File romDirectory;


	private static File changeListsDirectory;
	private static int offset;


	public static void openDefaultROM () {
		try {
			rom = new File(romDirectory + "/7th Saga Unedited.smc");
			bytes = Files.readAllBytes(rom.toPath());
			header = 0;

			Lists.CreateLists();
			Lists.InitializeLists();
			Data.serializeDefaultDataToDisk();
		} catch (Exception e) {
			System.out.println("Open default ROM failed");
		}


	}

	private static String removeFileExtension (File file) {
	String fileName = file.getName();
	int pos = fileName.lastIndexOf(".");
if(pos >0)

	{
		fileName = fileName.substring(0, pos);
	}
	return fileName;
}

    public static void openROM(File file)
    {
        rom = file;
		changeListsDirectory = new File(romDirectory + "/ChangeLists/" + removeFileExtension(rom));
        try
        {
            bytes = Files.readAllBytes(rom.toPath());

            int headerTest = ROM.bytes.length % 1024;
            switch (headerTest) {
                case 0:
                    ROM.header = 0;
                    break;
                case 512:
                    ROM.header = 512;
                    break;
                default:
                    ROM.header = 0;
                    System.out.println("improperly configured ROM");
                    break;
            }
            Data.serializeDefaultDataFromDisk();
            Lists.setState(new DataState());
            Lists.InitializeLists();


        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    public static void saveROM()
    {
        try
        {
            Lists.saveModels();
            Excel.makeChangeLists();
            Files.write(rom.toPath(), ROM.bytes, StandardOpenOption.WRITE);
            System.out.println("Data saved to Disk" + "\n");
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }


    public static File getProgramDirectory(){return programDirectory;}
    public static void setROMDirectory(File f){romDirectory=f;}
    public static File getROMDirectory(){return romDirectory;}
    public static void setProgramDirectory()
    {
        try
        {
            //programDirectory =new File(System.getProperty("java.class.path")).getParentFile();
            programDirectory = programDirectory = new File(".");
    }
        catch (Exception e)
        {
            programDirectory = new File(".");
        }
    }


    public static int getByte(int offset)
    {
        int unsignedByte= bytes[offset+header] & 0xFF;
        return unsignedByte;
    }

    // this returned value is little endian
    public static int getShort(int offset)
    {
        int unsignedByte1 = getByte(offset);
        int unsignedByte2 = getByte(offset+1);

        unsignedByte2<<=8;
        return unsignedByte1 | unsignedByte2;

    }

    // this returned value is little endian
    public static int getTriple(int offset)
    {
        int shortValue = getShort(offset);
        int finalByte = getByte(offset+2);
        finalByte<<=16;
        return finalByte|shortValue;
    }

    public static void setByte(int offset, int value)
    {
        bytes[offset+header]= (byte) value;
    }

    public static void setShort(int offset, int value)
    {
        setByte(offset, value&0xFF);
        value>>=8;
        setByte(offset+1, value&0xFF);
    }

    public static void setTriple(int offset, int value)
    {
        setShort(offset, value);
        value>>=16;
        setByte(offset+2, value&0xFF);
    }

    public static int getNextByte(){return getByte(offset++);}
    public static int getNextShort()
    {
        int value = getShort(offset);
        offset+=2;
        return value;
    }
    public static int getNextTriple(){
        int value = getTriple(offset);
        offset+=3;
        return value;

    }

    public static void setNextByte(int value){setByte(offset,value); offset++;}
    public static void setNextShort(int value){setShort(offset,value); offset+=2;}
    public static void setNextTriple(int value){setTriple(offset,value); offset+=3;}

    public static void setOffset (int pointer){offset=pointer;}
    public static int getOffset(){return offset;}

	public static File getChangeListsDirectory () {
		return changeListsDirectory;
	}




































}

File: src/application/ROM.kt
package application

import application.excel.*
import application.models.lists.Lists
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch
import java.io.BufferedInputStream
import java.io.File
import java.io.IOException
import java.io.ObjectInputStream
import java.nio.file.Files
import java.nio.file.StandardOpenOption
import kotlinx.coroutines.experimental.runBlocking

/**
 * Created by Seth on 4/3/2017.
 */
object ROM
{
	var bytes: ByteArray = byteArrayOf(1)
	private var header: Int = 0


	var programDirectory = File(".")
	private var rom = File(".")
	var romDirectory = File(".")
	var magicAndSpeedPatch = false


	var changeListsDirectory = File(".")
	var offset: Int = 0
	var settings = Settings()

	fun openDefaultROM()
	{
		try
		{
			rom = File(romDirectory.toString() + "/7th Saga Unedited.smc")
			bytes = Files.readAllBytes(rom.toPath())
			header = 0
			Lists.initializeDModels()
			val d = Data()
			d.serializeDefaultDataToDisk()
		}
		catch (e: Exception)
		{
			println("Open default ROM failed")
			e.printStackTrace()
		}
	}



	fun openROM(file: File)
	{
		rom = file
		changeListsDirectory = File(romDirectory.toString() + "/ChangeLists/" + removeFileExtension(rom))
		println(rom.canonicalPath)
		try
		{
			bytes = Files.readAllBytes(rom.toPath())

			val headerTest = ROM.bytes.size % 1024
			when (headerTest)
			{
				0    -> ROM.header = 0
				512  -> ROM.header = 512
				else ->
				{
					ROM.header = 0
					println("improperly configured ROM")
				}
			}

			// assemble Data object

			println("file path from disk: " + File(javaClass.getResource("models/defaultModels.data").toExternalForm()))
			val inputFile = BufferedInputStream(this.javaClass.getResourceAsStream("models/defaultModels.data"))
			// File defaultData = new File(Main.class.getResource("models/defaultModels.data").toURI());
			// System.out.println("File Path from disk: " + defaultData.getAbsolutePath());
			ObjectInputStream(inputFile).use {
				// deserialize the List
				val d = it.readObject() as Data
				d.serializeDefaultDataFromDisk()
			}

			Lists.initializeModels()
			if (getByte(0x7FFF0) and 1 > 0 && getByte(0x7FFF0) != 0xFF)
			{
				magicAndSpeedPatch = true
			}
		}
		catch(e: IOException)
		{
			e.printStackTrace()
		}
	}

	fun saveROM()
	{
		Lists.saveModels()
		println("Before save: " + Integer.toHexString(ROM.getTriple(0x6532)).toUpperCase())
		if(ROM.settings.createExcelFilesOnSave){saveToExcelFiles()}

		Files.write(rom.toPath(), ROM.bytes, StandardOpenOption.WRITE)
		println("Data saved to Disk" + "\n")
	}

	fun saveToExcelFiles()
	{
	launch()
	}

	fun standard()
	{
		val startTime = System.currentTimeMillis()

		CharacterComparisons("Character Comparisons.xlsx")
		CharacterData("Character.xlsx")
		EquipmentData("Equipment.xlsx")
		ExpData("Experience.xlsx")
		MonsterData("Monster.xlsx")
		val endTime = System.currentTimeMillis()
		val timeElapsed = (endTime - startTime)
		println("Excel Export Time is: ${timeElapsed.toDouble()/1000} seconds")
	}

	fun launch() = runBlocking {

		val startTime = System.currentTimeMillis()

		val job1 = launch{CharacterComparisons("Character Comparisons.xlsx")}
		val job2 = 	launch{	CharacterData("Character.xlsx")}
		val job3 = 	launch{EquipmentData("Equipment.xlsx")}
		val job4 = 	launch{ExpData("Experience.xlsx")}
		val job5 = 	launch{MonsterData("Monster.xlsx")}


		job1.join()
		job2.join()
		job3.join()
		job4.join()
		job5.join()
		val endTime = System.currentTimeMillis()
		val timeElapsed = (endTime - startTime)
		println("Excel Export Time is: ${timeElapsed.toDouble()/1000} seconds")
	}

	fun acyncExcel()
	{
		val startTime = System.currentTimeMillis()

		val cc = async{CharacterComparisons("Character Comparisons.xlsx")}
		val ch = async{CharacterData("Character.xlsx")}
		val eq = async{EquipmentData("Equipment.xlsx")}
		val ex = async{ExpData("Experience.xlsx")}
		val mo = async{MonsterData("Monster.xlsx")}

		runBlocking {
			cc.await()
			ch.await()
			eq.await()
			ex.await()
			mo.await()
			val endTime = System.currentTimeMillis()
			val timeElapsed = (endTime - startTime)
			println("Excel Export Time is: ${timeElapsed.toDouble()/1000} seconds")
		}
	}





	fun setProgramDirectory()
	{
		try
		{
			//programDirectory =new File(System.getProperty("java.class.path")).getParentFile();
			programDirectory = File(".")
			programDirectory = programDirectory
		}
		catch (e: Exception)
		{
			programDirectory = File(".")
		}

	}

	private fun removeFileExtension(file: File): String
	{
		var fileName = file.name
		val pos = fileName.lastIndexOf(".")
		if (pos > 0) fileName = fileName.substring(0, pos)
		return fileName
	}

	fun getByte(offset: Int): Int
	{
		val unsignedByte: Int = bytes[offset + header].toInt() and 0xFF
		return unsignedByte
	}

	// this returned value is little endian
	fun getShort(offset: Int): Int
	{
		val unsignedByte1 = getByte(offset)
		var unsignedByte2 = getByte(offset + 1)

		unsignedByte2 = unsignedByte2 shl 8
		return unsignedByte1 or unsignedByte2

	}

	// this returned value is little endian
	fun getTriple(offset: Int): Int
	{
		val shortValue = getShort(offset)
		var finalByte = getByte(offset + 2)
		finalByte = finalByte shl 16
		return finalByte or shortValue
	}

	fun setByte(offset: Int, value: Int)
	{
		bytes[offset + header] = value.toByte()
	}

	fun setShort(offset: Int, value: Int)
	{
		var value = value
		setByte(offset, value and 0xFF)
		value = value shr 8
		setByte(offset + 1, value and 0xFF)
	}

	fun setTriple(offset: Int, value: Int)
	{
		var value = value
		setShort(offset, value)
		value = value shr 16
		setByte(offset + 2, value and 0xFF)
	}

	var nextByte: Int
		get() = getByte(offset++)
		set(value)
		{
			setByte(offset, value)
			offset++
		}
	var nextShort: Int
		get()
		{
			val value = getShort(offset)
			offset += 2
			return value
		}
		set(value)
		{
			setShort(offset, value)
			offset += 2
		}
	var nextTriple: Int
		get()
		{
			val value = getTriple(offset)
			offset += 3
			return value
		}
		set(value)
		{
			setTriple(offset, value)
			offset += 3
		}
}
